File: pdb2fasta.R
#!/usr/bin/env Rscript
args=commandArgs(TRUE)
if (length(args)<1){
    cat("pdb2fasta.py pdb.pdb > seq.fasta\n")
    cat("    convert PDB file pdb.pdb to FASTA sequence file seq.fasta\n")
}

aa3to1=c("A","V","F","P","M","I","L","D","E","K",
     "R","S","T","Y","H","C","N","Q","W","G","M")
names(aa3to1)=c("ALA","VAL","PHE","PRO","MET","ILE","LEU","ASP","GLU","LYS",
          "ARG","SER","THR","TYR","HIS","CYS","ASN","GLN","TRP","GLY","MSE")

for (pdb_file in args){
    chain_prev="" # chain ID of previous chain
    filename=tools::file_path_sans_ext(basename(pdb_file))
    fp=file(pdb_file,'r')
    while(1){
        line=readLines(fp,1)
        if (length(line)<1||substr(line,1,6)=="ENDMDL"){break}
        chainID=substr(line,22,22)
        resn=substr(line,18,20)
        atom=substr(line,13,16)
	altid=substr(line,17,17)
        if ((substr(line,1,6)!="ATOM  " && substr(line,1,6)!="HETATM")
           ||atom!=" CA "||is.element(resn,names(aa3to1))==FALSE
	   ||(altid!=" " && altid!="A")){next} # continue
        if (chainID !=chain_prev){
            if (chain_prev==""){
                cat(paste(">",filename,":",chainID,"\n",sep=""))
            }else{
                cat(paste("\n>",filename,":",chainID,"\n",sep=""))
            }
            chain_prev=chainID
        }
        cat(aa3to1[resn])
    }
    close(fp)
    cat("\n")
}

File: pdb2fasta.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.*;
import java.io.*;

public class pdb2fasta 
{
    public static void main(String[] args)
    {
        if (args.length==0)
        {
            System.out.println("java pdb2fasta pdb.pdb > seq.fasta\n" + 
            "    convert PDB file pdb.pdb to FASTA sequence seq.fasta\n");
            return;
        }
        String[] aa3 = {"ALA", "VAL", "PHE", "PRO", "MET", "ILE", "LEU",
                        "ASP", "GLU", "LYS", "ARG", "SER", "THR", "TYR",
                        "HIS", "CYS", "ASN", "GLN", "TRP", "GLY", "MSE"};
        String[] aa1 = {"A", "V", "F", "P", "M", "I", "L", "D", "E",
                        "K", "R", "S", "T", "Y", "H", "C", "N", "Q",
                        "W", "G", "M"};
        Map<String, String> aa3to1=new HashMap();
        for(int i = 0; i < aa3.length; i++) aa3to1.put(aa3[i], aa1[i]);
        Pattern ca_pattern=Pattern.compile(
            "^ATOM\\s{2,6}\\d{1,5}\\s{2}CA\\s[\\sA]([A-Z]{3})\\s([\\s\\w])"+ 
            "|^HETATM\\s{0,4}\\d{1,5}\\s{2}CA\\s[\\sA](MSE)\\s([\\s\\w])");

        for (int arg=0;arg<args.length;arg++)
        {
            File pdb_file = new File(args[arg]);
            String filename=pdb_file.getName();
            if (filename.indexOf(".")>-1)
                filename=filename.substring(0,filename.indexOf("."));
            Map chain_dict = new HashMap();
            List chain_list = new ArrayList();
            String line;
            String resn;
            String chain;
            try
            {
                BufferedReader fp = new BufferedReader(new FileReader(pdb_file));
                while( (line = fp.readLine()) != null &&
                    !line.startsWith("ENDMDL")) 
                {
                    Matcher match_list = ca_pattern.matcher(line);
                    if(match_list.find())
                    {
                        resn = match_list.group(1);
                        chain = match_list.group(2);
                        if (chain_dict.containsKey(chain))
                        {
                            chain_dict.put(chain, chain_dict.get(chain)+
                                                  aa3to1.get(resn));
                        }
                        else
                        {
                            chain_dict.put(chain, aa3to1.get(resn));
                            chain_list.add(chain);
                        }
                    }
                }
                fp.close();
           }
           catch(FileNotFoundException ex) {ex.printStackTrace();}
           catch(IOException ex) {ex.printStackTrace();}
           for (int c = 0; c < chain_list.size(); c++)
               System.out.println(">"+filename+":"+chain_list.get(c)+"\n"+
                   chain_dict.get(chain_list.get(c)));
        }
    }
}
